compiler/fir/checkers/gen/org/jetbrains/kotlin/fir/analysis/diagnostics/FirErrors.kt

```diff
@@ -104,6 +105,9 @@ object FirErrors {
     val PRIMARY_CONSTRUCTOR_REQUIRED_FOR_DATA_CLASS by warning0<FirSourceElement, PsiElement>()
     val EXPLICIT_DELEGATION_CALL_REQUIRED by warning0<FirSourceElement, PsiElement>(SourceElementPositioningStrategies.SECONDARY_CONSTRUCTOR_DELEGATION_CALL)
     val SEALED_CLASS_CONSTRUCTOR_CALL by error0<FirSourceElement, PsiElement>()
+    val DATA_CLASS_WITHOUT_PARAMETERS by error0<FirSourceElement, KtPrimaryConstructor>()
```

Comment:
This should be just `PsiElement` instead of `KtPrimaryConstructor`?

Reply:
I assume this is why the test fails like:
```
  Caused by: java.lang.RuntimeException: com.intellij.testFramework.LoggedErrorProcessor$TestLoggerAssertionError: org.jetbrains.kotlin.psi.KtClass cannot be cast to org.jetbrains.kotlin.psi.KtPrimaryConstructor
```

Reply:
Thanks! Actually I am not sure why we specify a type here.

Reply:
This type parameter is eventually passed to: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/SourceElementPositioningStrategy.kt#L14 and then used by reported position/element's validation check. That's why we try to match old FE's diagnostics type.

Reply:
It is better to have a concrete psi type in diagnostics. This will help writing IDE quickfixes by reducing type checks boilerplate

Reply:
It's not clear for me why we have `KtPrimaryConstructor` here and `PsiElement` in `FirDiagnosticList`, probably you haven't regenerated something... Anyway, I'd suggest to use concrete psi type; probably `KtPrimaryConstructor` is not possible here because it may not exist in cases like `data class MyDataClass`. May be we could use `KtClass` or `KtClassOrObject`? Please consider this way. cc @tgeng. @darthorimar, what do you think about it?

Reply:
I'm for `KtClass` as the more concrete (`KtClassOrObject` is not applicable since object can not have constructors)

Reply:
I think using `KtPrimaryConstructor` is appropriate. If there are no constructor, `PRIMARY_CONSTRUCTOR_REQUIRED_FOR_DATA_CLASS` should be reported.

Reply:
Ah now I remember why the test failed. Even if no primary constructor is provided, FIR still creates `FirPrimaryConstructor`. Since in this case we should be reporting `PRIMARY_CONSTRUCTOR_REQUIRED_FOR_DATA_CLASS`. I should really suppress reporting of `DATA_CLASS_WITHOUT_PARAMETERS` in the checker so that the class cast exception won't happen in the first place.

Reply:
Also I am not sure why `PRIMARY_CONSTRUCTOR_REQUIRED_FOR_DATA_CLASS` is marked a warning in FIR. It's an error in FE1.0.